core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2262,21 +2262,39 @@ private[spark] object Utils extends Logging {
   }
 
   /**
-   * Return whether dynamic allocation is enabled in the given conf
-   * Dynamic allocation and explicitly setting the number of executors are inherently
-   * incompatible. In environments where dynamic allocation is turned on by default,
-   * the latter should override the former (SPARK-9092).
+   * Return whether dynamic allocation is enabled in the given conf.
    */
   def isDynamicAllocationEnabled(conf: SparkConf): Boolean = {
-    val numExecutor = conf.getInt("spark.executor.instances", 0)
     val dynamicAllocationEnabled = conf.getBoolean("spark.dynamicAllocation.enabled", false)
-    if (numExecutor != 0 && dynamicAllocationEnabled) {
-      logWarning("Dynamic Allocation and num executors both set, thus dynamic allocation disabled.")
-    }
-    numExecutor == 0 && dynamicAllocationEnabled &&
+    dynamicAllocationEnabled &&
       (!isLocalMaster(conf) || conf.getBoolean("spark.dynamicAllocation.testing", false))
   }
 
+  /**
+   * Return the minimum number of executors for dynamic allocation.
+   */
+  def getDynamicAllocationMinExecutors(conf: SparkConf): Int = {
+    conf.getInt("spark.dynamicAllocation.minExecutors", 0)
+  }
+
+  /**
+   * Return the maximum number of executors for dynamic allocation.
+   */
+  def getDynamicAllocationMaxExecutors(conf: SparkConf): Int = {
+    conf.getInt("spark.dynamicAllocation.maxExecutors", Integer.MAX_VALUE)
+  }
+
+  /**
+   * Return the initial number of executors for dynamic allocation.
+   */
+  def getDynamicAllocationInitialExecutors(conf: SparkConf): Int = {
+    math.max(
+      math.max(
+        conf.getInt("spark.dynamicAllocation.initialExecutors", 0),
```

Comment:
I don't think the config API can handle both calls to `math.max` here... you could replace the inner one with a fallback config (`initialExecutors` falls back to the value of `spark.executor.instances` if not set), although even then the semantics are slightly different.

In any case, this might be more readable as:

```
Seq(
  conf.getInt("spark.dynamicAllocation.initialExecutors", 0),
  conf.getInt("spark.executor.instances", 0),
  getDynamicAllocationMinExecutors(conf)).max
```

Assuming I got the semantics right.